cpp/src/arrow/compute/kernels/compare.cc

```diff
@@ -17,250 +17,296 @@
 
 #include "arrow/compute/kernels/compare.h"
 
+#include <utility>
+
+#include "boost/range.hpp"
+
 #include "arrow/compute/context.h"
 #include "arrow/compute/kernel.h"
 #include "arrow/compute/kernels/util_internal.h"
 #include "arrow/util/bit_util.h"
+#include "arrow/util/checked_cast.h"
 #include "arrow/util/logging.h"
+#include "arrow/util/string_view.h"
+#include "arrow/visitor_inline.h"
 
 namespace arrow {
 
+using internal::checked_cast;
+using internal::checked_pointer_cast;
+using util::string_view;
+
 namespace compute {
 
-std::shared_ptr<DataType> CompareBinaryKernel::out_type() const {
-  return compare_function_->out_type();
-}
+template <typename T, CompareOperator Op>
+struct Comparator;
+
+template <typename T>
+struct Comparator<T, CompareOperator::EQUAL> {
+  constexpr static bool Compare(const T& lhs, const T& rhs) { return lhs == rhs; }
+};
+
+template <typename T>
+struct Comparator<T, CompareOperator::NOT_EQUAL> {
+  constexpr static bool Compare(const T& lhs, const T& rhs) { return lhs != rhs; }
+};
+
+template <typename T>
+struct Comparator<T, CompareOperator::GREATER> {
+  constexpr static bool Compare(const T& lhs, const T& rhs) { return lhs > rhs; }
+};
+
+template <typename T>
+struct Comparator<T, CompareOperator::GREATER_EQUAL> {
+  constexpr static bool Compare(const T& lhs, const T& rhs) { return lhs >= rhs; }
+};
+
+template <typename T>
+struct Comparator<T, CompareOperator::LESS> {
+  constexpr static bool Compare(const T& lhs, const T& rhs) { return lhs < rhs; }
+};
+
+template <typename T>
+struct Comparator<T, CompareOperator::LESS_EQUAL> {
+  constexpr static bool Compare(const T& lhs, const T& rhs) { return lhs <= rhs; }
+};
+
+template <typename Value>
```

Comment:
I suppose those ranges will be useful for other kernels at some point?

Reply:
Definitely. As a (much) larger project I plan to eventually extract these and other range impls, consolidating into `arrow/util/range.h`.